blockprod/src/detail/mod.rs

```diff
@@ -379,9 +376,9 @@ impl BlockProduction {
     fn spawn_block_solver(
         &self,
         current_tip_index: &GenBlockIndex,
-        stop_flag: Arc<AtomicBool>,
+        stop_flag: Arc<RelAtomicBool>,
         block_body: &BlockBody,
-        block_timestamp_seconds: Arc<AtomicU64>,
+        block_timestamp_seconds: Arc<RelAtomicU64>,
```

Comment:
This has to be `SeqCst` or at least `AcqRel`.

Reply:
I don't think it's really needed, because there is no synchronization on `block_timestamp_seconds` AKA `last_timestamp_seconds_used`. The actual synchronization points should be the call of `tokio::select!` (the `result_receiver` part) in `BlockProduction::produce_block_with_custom_id` and `result_sender.send(...)` in `BlockProduction::spawn_block_solver`. So the real synchronization is actually encapsulated inside `BlockProduction`, which is a good thing. And I would argue that making `block_timestamp_seconds/last_timestamp_seconds_used` a "SynAtomic" would introduce code smell, because it will suggest that we have a low-level synchronization primitive that is a part of a package's public interface.
But I definitely understand your concerns, because even if it's not used for synchronization, this fact is still very non-obvious. So, I don't argue about making that parameter a SynAtomic, but I also want at least to write a TODO to remove it completely, preferably ASAP. Ideally, Pavel should remove it because he added it relatively recently, but if he's busy and there are no objections, I could try to do it myself (but I'll have to ask some stupid questions in the process).

Reply:
The problem, because of which I think we should use SynAtomic, is that the main thread (where `produce_block_with_custom_id` resides) should learn that the atomic value has been updated in the new thread. I would rather use an atomic fence, but it's not worth it. What I fear is, through the lack of synchronization, what will happen is that the `spawn_block_solver` will update the value of `last_timestamp_seconds_used`, and then the main thread won't notice due to the lack of enforcement of cache-reloads. The current implementation doesn't seem to take into account that we have to continue staking from where we left of, so `last_timestamp_seconds_used` should probably be updated in the main thread and returned to the request sender (rpc, for example). But again, one problem at a time. 

Like I said before: The fear is always that someone looks at the code at face value and changes something without us noticing. So making this SynAtomic is a good guarantee for future changes. Feel free to add a TODO for when we're done with everything related, to check this problem again.

Reply:
Just to be clear: I don't really care if the synchronization will happen because some other part/primitive of the code has sufficient synchronization semantics. This feels to me like playing with fire and creates complex interdependencies, especially that depending on the program, it may be in another memory page, which may or may not synchronize. Having worked with emulators, I've seen how complex the memory model in CPUs is compared to the naive all-or-nothing mental model we usually have in mind. Yet I still assert my ignorance in this topic and I'll never learn enough about it. 

Having said that, `stop_flag` is a whole other discussion. Like I mentioned before, `stop_flag` is allowed to have all the delay it needs because the worst thing that can happen is a delay in response. On the other hand, lack of synchronization for `last_timestamp_seconds_used` can lead to incorrect value in the main thread. Almost impossible to see in real life, thanks to x86_64's toleraly AcqRel for everything, but still.

I really don't like these discussions to happen in text. It's a waste of time, leads to misunderstandings and mystifies nuance and given that it doesn't affect performance to use the more synchronizing option, I would rather pick whatever works now and move on. If you wish, let's have a call to have that discussion so that we don't spend 30 minutes writing text that will be buries in less than a week. Let's have a call whenever you wish.

Reply:
We came to the conclusion that we'll return the timestamp with the error and not use atomic. So the value will be moved back and forth instead of using an atomic.